File: strains/plotruntimes.py
#!/usr/bin/env python

"""
Script to generate a plot of observing run times of various interferometric gravitational wave detectors
for the initial detector era (TAMA300, GEO600, LIGO and Virgo).

Copyright: Matthew Pitkin, 2016
"""

from __future__ import division

from collections import OrderedDict as orddic

import numpy as np
import matplotlib as mpl
from matplotlib import pyplot as pl

mplparams = {
      'backend': 'Agg',
      'text.usetex': True, # use LaTeX for all text
      'axes.linewidth': 0.5, # set axes linewidths to 0.5
      'axes.grid': True, # add a grid
      'grid.linewidth': 0.5,
      'grid.alpha': 0.3,
      'font.family': 'sans-serif',
      'font.sans-serif': 'Avant Garde, Helvetica, Computer Modern Sans serif',
      'font.size': 18 }

mpl.rcParams.update(mplparams)

# offsets for text run label placement
runyoffsetbelow = -0.4
runyoffsetabove = 0.2

# set LIGO run times [start end] (years)
LIGO = {'S1': {'times': [(2002+(7/12)+(23/365.25)), (2002+(8/12)+(9/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'S2': {'times': [(2003+(1/12)+(14/365.25)), (2003+(3/12)+(14/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'S3': {'times': [(2003+(9/12)+(31/365.25)), (2004+(9/365.25))],         'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'S4': {'times': [(2005+(1/12)+(22/365.25)), (2005+(2/12)+(23/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'S5': {'times': [(2005+(10/12)+(4/365.25)), (2007+(9/12))],             'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'Astrowatch': {'times': [(2008 + (1/12) + (7/365.25)), (2009 + (4/12) + (28/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': None, 'alpha': 0.45}},
        'S6': {'times': [(2009+(6/12)+(7/365.25)), (2010+(9/12)+(20/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}}}

# GEO600 science runs (S1-6)
GEO = {'S1':   {'times': LIGO['S1']['times'], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
       'S3I':  {'times': [(2003+(10/12)+(5/365.25)), (2003+(10/12)+(11/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
       'S3II': {'times': [(2003+(11/12)+(30/365.25)), (2004+(13/365.25))], 'textoffset': runyoffsetabove, 'kwargs': {'marker': 'o'}},
       'S4':   {'times': LIGO['S4']['times'], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
       'preS5':{'times': [(2006+20/365.25), (2006+(4/12))], 'textoffset': runyoffsetbelow*1e6, 'kwargs': {'marker': None, 'alpha': 0.45}}, # pre-S5 overnight and weekend astrowatch
       'S5':   {'times': [(2006+(4/12)), (2006+(9/12)+(15/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
       'Astrowatch': {'times': [(2006+(9/12)+(15/365.25)), (2009+(6/12)+(7/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': None, 'alpha': 0.45}}, # to start of S6
       'S6/Astrowatch': {'times': [(2009 + (6/12) + (7/365.25)), 2012], 'textoffset': runyoffsetbelow, 'kwargs': {'alpha': 0.45}}}

# TAMA data taking runs (DT1-9)
TAMA = {'DT1': {'times': [(1999+(7/12)+(6/365.25)),   (1999+(7/12)+(7/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT2': {'times': [(1999+(8/12)+(17/365.25)),  (1999+(8/12)+(20/365.25))], 'textoffset': runyoffsetabove, 'kwargs': {'marker': 'o'}},
        'DT3': {'times': [(2000+(3/12)+(20/365.25)),  (2000+(3/12)+(23/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT4': {'times': [(2000+(7/12)+(21/365.25)),  (2000+(8/12)+(4/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT5': {'times': [(2001+(2/12)+(2/365.25)),   (2001+(2/12)+(8/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT6': {'times': [(2001+(7/12)+(0/365.25)),   (2001+(8/12)+(20/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT7': {'times': [(2002+(7/12)+(31/365.25)),  (2002+(8/12)+(2/365.25))],  'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT8': {'times': LIGO['S2']['times'], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
        'DT9': {'times': [(2003+(10/12)+(28/365.25)), (2004+10/365.25)], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}}}

# Virgo runs, weekend science run (WSR) and VSR1-4
VIRGO = {'WSR1': {'times': [(2006+(8/12)+(8/365.25)), (2007+(2/12)+(12/365.25))], 'textoffset': runyoffsetbelow*1e6, 'kwargs': {'marker': None, 'alpha': 0.45}},
         'VSR1': {'times': [(2007+(4/12)+(18/365.25)), (2007+(9/12))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
         'A5':   {'times': [(2007+(9/12)), (2009+(6/12)+(7/365.25))], 'textoffset': runyoffsetbelow*1e6, 'kwargs': {'marker': None, 'alpha': 0.45}},
         'VSR2': {'times': [(2009+(6/12)+(7/365.25)), (2010+(8/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
         'VSR3': {'times': [(2010+(7/12)+(11/365.25)), (2010+(6/12)+(20/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}},
         'VSR4': {'times': [(2011+(5/12)+(3/365.25)), (2011+(8/12)+(5/365.25))], 'textoffset': runyoffsetbelow, 'kwargs': {'marker': 'o'}}}

# create figure
fig = pl.figure(figsize=(18, 5))

# get axes
ax = pl.gca()

# draw some thin horozinatal lines for asthetics
ax.plot([1999, 2012], [1, 1], 'k--', linewidth=0.5)
ax.plot([1999, 2012], [2, 2], 'k--', linewidth=0.5)
ax.plot([1999, 2012], [3, 3], 'k--', linewidth=0.5)
ax.plot([1999, 2012], [4, 4], 'k--', linewidth=0.5)

# detector positions in plot
detpos = orddic([('LIGO', 2), ('GEO600', 3), ('TAMA300',  1), ('Virgo', 4)])

# plot LIGO runs
for run in LIGO:
  LIGO[run]['kwargs']['color'] = 'm'
  LIGO[run]['kwargs']['linewidth'] = 6
  LIGO[run]['kwargs']['markersize'] = 10
  LIGO[run]['kwargs']['markeredgecolor'] = 'none'
  ax.plot(LIGO[run]['times'], [detpos['LIGO'], detpos['LIGO']], **LIGO[run]['kwargs'])
  ax.text(LIGO[run]['times'][0] + 0.5*np.diff(LIGO[run]['times']), detpos['LIGO'] + LIGO[run]['textoffset'], run, fontsize=12, ha='center')

# plot GEO600 runs
for run in GEO:
  GEO[run]['kwargs']['color'] = 'b'
  GEO[run]['kwargs']['linewidth'] = 6
  GEO[run]['kwargs']['markersize'] = 10
  GEO[run]['kwargs']['markeredgecolor'] = 'none'
  ax.plot(GEO[run]['times'], [detpos['GEO600'], detpos['GEO600']], **GEO[run]['kwargs'])
  ax.text(GEO[run]['times'][0] + 0.5*np.diff(GEO[run]['times']), detpos['GEO600'] + GEO[run]['textoffset'], run, fontsize=12, ha='center')

# plot TAMA300 runs and labels
for run in TAMA:
  TAMA[run]['kwargs']['color'] = 'r'
  TAMA[run]['kwargs']['linewidth'] = 6
  TAMA[run]['kwargs']['markersize'] = 10
  TAMA[run]['kwargs']['markeredgecolor'] = 'none'
  ax.plot(TAMA[run]['times'], [detpos['TAMA300'], detpos['TAMA300']], **TAMA[run]['kwargs'])
  ax.text(TAMA[run]['times'][0] + 0.5*np.diff(TAMA[run]['times']), detpos['TAMA300'] + TAMA[run]['textoffset'], run, fontsize=12, ha='center')

# plot Virgo runs
for run in VIRGO:
  VIRGO[run]['kwargs']['color'] = 'g'
  VIRGO[run]['kwargs']['linewidth'] = 6
  VIRGO[run]['kwargs']['markersize'] = 10
  VIRGO[run]['kwargs']['markeredgecolor'] = 'none'
  ax.plot(VIRGO[run]['times'], [detpos['Virgo'], detpos['Virgo']], **VIRGO[run]['kwargs'])
  ax.text(VIRGO[run]['times'][0] + 0.5*np.diff(VIRGO[run]['times']), detpos['Virgo'] + VIRGO[run]['textoffset'], run, fontsize=12, ha='center')

# add other LIGO annotation
# first lock of H2 (not actual date it happened, but when it was announced https://labcit.ligo.caltech.edu/LIGO_web/firstlock/)
ax.annotate('first lock H2', xy=(2000+(9/12)+(20/365.25), detpos['LIGO']), xytext=(2000+(9/12)+(20/365.25), detpos['LIGO']-0.5), fontsize=12,
            arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), ha='center')
# full lock all detectors
ax.annotate('full lock\nall detectors', xy=(2002.25, detpos['LIGO']), xytext=(2002.25, detpos['LIGO']-0.5), fontsize=12,
            arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), ha='center', va='center', multialignment='center')
# Note Enhanced LIGO period
ax.text(LIGO['S6']['times'][0] + 0.5*np.diff(LIGO['S6']['times']), detpos['LIGO'] + runyoffsetabove,
        'Enhanced LIGO', fontsize=12, ha='center')

# add other GEO annotation
# first lock of power recycled Michelson (25th Oct 2001 from Appendix E of Hartmut's thesis http://inspirehep.net/record/1228030/files/Grote,%20Hartmut_Diss.pdf)
ax.annotate('first lock of full\npower-recycled Michelson', xy=(2001+(9/12)+(25/365.25), detpos['GEO600']), xytext=(2001+(9/12)+(25/365.25), detpos['GEO600']-0.5), fontsize=12,
            arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), ha='center', va='center', multialignment='center')
# first lock of dual-recycled Michelson (15 Nov 2002 from Appendix E of Hartmut's thesis http://inspirehep.net/record/1228030/files/Grote,%20Hartmut_Diss.pdf)
ax.annotate('first lock of\ndual-recycled Michelson', xy=(2002+(10/12)+(15/365.25), detpos['GEO600']), xytext=(2002+(10/12)+(15/365.25), detpos['GEO600']+0.5), fontsize=12,
            arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), ha='center', va='center')
# Note the overnight and weekend run period
ax.annotate('Overnight and\nweekend', xy=(GEO['S5']['times'][0] - 2./12, detpos['GEO600']), xytext=(GEO['S5']['times'][0], detpos['GEO600']+0.3), textcoords='data', multialignment='center', arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), fontsize=12)
ax.annotate('Overnight and\nweekend', xy=(GEO['S5']['times'][1] + 8./12, detpos['GEO600']), xytext=(GEO['S5']['times'][0], detpos['GEO600']+0.3), xycoords='data', textcoords='data', multialignment='center', arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), fontsize=12, alpha=0.0)
# Note GEO-HF period
ax.text(GEO['S6/Astrowatch']['times'][0] + 0.5*np.diff(GEO['S6/Astrowatch']['times']), detpos['GEO600'] + runyoffsetabove,
        'GEO-HF', fontsize=12, ha='center')

# add other Virgo annotation
# first full lock of IFO with power-recycling 26 Oct 2004 http://iopscience.iop.org/article/10.1088/0264-9381/22/18/S01/meta
ax.annotate('first full lock with\npower-recycling', xy=(2004+(9/12)+(26/365.25), detpos['Virgo']), xytext=(2004+(9/12)+(26/365.25), detpos['Virgo']+0.5), fontsize=12,
            arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), ha='center', va='center', multialignment='center')
# add weekend science run label
ax.annotate('Weekend\nScience Run', xy=(VIRGO['VSR1']['times'][0] - 6./12, detpos['Virgo']), xytext=(VIRGO['VSR1']['times'][0]-3/12, detpos['Virgo']+0.3), textcoords='data', multialignment='center', arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), fontsize=12)
ax.annotate('Weekend\nScience Run', xy=(VIRGO['VSR1']['times'][1] + 6./12, detpos['Virgo']), xytext=(VIRGO['VSR1']['times'][0]-3/12, detpos['Virgo']+0.3), xycoords='data', textcoords='data', multialignment='center', arrowprops=dict(facecolor='black', shrink=0.05, width=1, headwidth=6, headlength=5), fontsize=12, alpha=0.0)
# add Virgo+
ax.text(VIRGO['VSR3']['times'][0] + 0.5*np.diff(VIRGO['VSR3']['times']), detpos['Virgo'] + runyoffsetabove,
        'Virgo+', fontsize=12, ha='center')

# set plot limits
pl.xlim([1999.25, 2012])
pl.ylim([0, 5])

pl.xlabel('Year')

# set y-axis labels to detector names
ax.set_yticks([1, 2, 3, 4])
ax.set_yticklabels([detpos.keys()[detpos.values().index(v)] for v in sorted(detpos.values())])

# set x-axis ticks and labels
xticks = np.arange(1999.25, 2012.1, 0.25).tolist()
xticklabels = ['' for i in range(len(xticks))]
for i, v in enumerate(xticks):
  if v%1 == 0.:
    xticklabels[i] = str(int(v))
ax.set_xticks(xticks)
ax.set_xticklabels(xticklabels)

pl.tight_layout()

fig.savefig('../figures/runtimes/runtimes.png', dpi=400)
fig.savefig('../figures/runtimes/runtimes.pdf')

# rasterize for eps output with transparencies
ax.set_rasterized(True)
fig.savefig('../figures/runtimes/runtimes.eps', dpi=400)

#pl.show()

File: strains/plotruntimes.m
% a script to create a timeline of LIGO, GEO, VIRGO and TAMA science runs

% edit to add extra runs

% LIGO science runs (S1-6) [start end] (in years)
LIGOS1 = [(2002+(7/12)+(23/365.25)) (2002+(8/12)+(9/365.25))];
LIGOS2 = [(2003+(1/12)+(14/365.25)) (2003+(3/12)+(14/365.25))];
LIGOS3 = [(2003 + (9/12)+(31/365.25)) (2004+(9/365.25))];
LIGOS4 = [(2005 + (1/12) + (22/365.25)) (2005 + (2/12) + (23/365.25))];
LIGOS5 = [(2005 + (10/12) + (4/365.25)) (2007 + (9/12))];
LIGOA5 = [(2008 + (1/12) + (7/365.25)) (2009 + (4/12) + (28/365.25))]; % astrowatch to start of S6
LIGOS6 = [(2009 + (6/12) + (7/365.25)) (2010 + (9/12) + (20/365.25))];

% GEO600 science runs (S1-6)
GEOS1 = LIGOS1;
GEOS3I = [(2003 + (10/12) + (5/365.25)) (2003 + (10/12) + (11/365.25))];
GEOS3II = [(2003 + (11/12) + (30/365.25)) (2004 + (13/365.25))];
GEOS4 = LIGOS4;
GEOS5start = [(2006 + 20/365.25) (2006 + (4/12))];
GEOS5247 = [(2006 + (4/12)) (2006 + (9/12) + (15/365.25))];
GEOA5 = [(2006 + (9/12) + (15/365.25)) (2009 + (6/12) + (7/365.25))]; % currently to start of S6
GEOS6 = [(2009 + (6/12) + (7/365.25)) 2012];

% TAMA data taking runs (DT1-9)
TAMADT1 = [(1999 + (7/12) + (6/365.25)) (1999 + (7/12) + (7/365.25))];
TAMADT2 = [(1999 + (8/12) + (17/365.25)) (1999 + (8/12) + (20/365.25))];
TAMADT3 = [(2000 + (3/12) + (20/365.25)) (2000 + (3/12) + (23/365.25))];
TAMADT4 = [(2000 + (7/12) + (21/365.25)) (2000 + (8/12) + (4/365.25))];
TAMADT5 = [(2001 + (2/12) + (2/365.25)) (2001 + (2/12) + (8/365.25))];
TAMADT6 = [(2001 + (7/12) + (0/365.25)) (2001 + (8/12) + (20/365.25))];
TAMADT7 = [(2002 + (7/12) + (31/365.25)) (2002 + (8/12) + (2/365.25))];
TAMADT8 = LIGOS2;
TAMADT9 = [(2003 + (10/12) + (28/365.25)) (2004 + 10/365.25)];

% VIRGO runs, weekend science run (WSR) and VSR1-4
VIRGOWSR = [(2006 + (8/12) + (8/365.25)) (2007 + (2/12) + (12/365.25))];
VIRGOVSR1 = [(2007 + (4/12) + (18/365.25)) (2007+(9/12))];
VIRGOA5 = [(2007+(9/12)) (2009 + (6/12) + (7/365.25))];
VIRGOVSR2 = [(2009 + (6/12) + (7/365.25)) (2010 + (8/365.25))];
VIRGOVSR3 = [(2010 + (7/12) + (11/365.25)) (2010 + (6/12) + (20/365.25))];
VIRGOVSR4 = [(2011 + (5/12) + (3/365.25)) (2011 + (8/12) + (5/365.25))]; 

% make plot (currently goes to 2012)
% draw some thin horozinatal lines for asthetics 
plot([1999 2012], [1 1], 'k--', 'linewidth', 0.5)
hold on
plot([1999 2012], [2 2], 'k--', 'linewidth', 0.5)
plot([1999 2012], [3 3], 'k--', 'linewidth', 0.5)
plot([1999 2012], [4 4], 'k--', 'linewidth', 0.5)

% plot LIGO
plot(LIGOS1, [1 1], 'm-o', LIGOS2, [1 1], 'm-o', LIGOS3, [1 1], 'm-o',...
    LIGOS4, [1 1], 'm-o', LIGOS5, [1 1], 'm-o', LIGOA5, [1 1], 'm--', ...
    LIGOS6, [1 1], 'm-o', 'linewidth', 3, 'markersize', 4);

% plot GEO
plot(GEOS1, [2 2], 'b-o', GEOS3I, [2 2], 'b-o', GEOS3II, [2 2], 'b-o',...
    GEOS4, [2 2], 'b-o', GEOS5start, [2 2], 'b--', GEOS5247, [2 2], ...
    'b-o', GEOA5, [2 2], 'b--o', GEOS6, [2 2], 'b-->', 'linewidth', 3, ...
    'markersize', 4);

% plot TAMA
plot(TAMADT1, [3 3], 'r-o', TAMADT2, [3 3], 'r-o',TAMADT3, [3 3], ...
    'r-o',TAMADT4, [3 3], 'r-o',TAMADT5, [3 3], 'r-o',TAMADT6, [3 3], ...
    'r-o',TAMADT7, [3 3], 'r-o',TAMADT8, [3 3], 'r-o',TAMADT9, [3 3], ...
    'r-o', 'linewidth', 3, 'markersize', 4);

% plot VIRGO
plot(VIRGOWSR, [4 4], 'g--', VIRGOVSR1, [4 4], 'g-o', ...
    VIRGOA5, [4 4], 'g--', ...
    VIRGOVSR2, [4 4], 'g-o', VIRGOVSR3, [4 4], 'g-o', ...
    VIRGOVSR4, [4 4], 'g-o', ...
    'linewidth', 3, 'markersize', 4);

% set y axis (0-5 to give a space at the top and bottom)
ylim([0 5]);

% add detector names to y axis
set(gca, 'ytick', [0 1 2 3 4 5], ...
    'yticklabel', {'', 'LIGO', 'GEO600', 'TAMA300', 'VIRGO', ''});

% set the x axis from 1999.5 to 2012 (can change)
xlim([1999.5 2012]);

% add tick marks every quarter year, but only label the years
set(gca, 'xtick', [1999.5 1999.75 2000 2000.25 2000.5 2000.75 2001 ...
    2001.25 2001.5 2001.75 2002 2002.25 2002.5 2002.75 2003 2003.25 ...
    2003.5 2003.75 2004 2004.25 2004.5 2004.75 2005 2005.25 2005.5 ...
    2005.75 2006 2006.25 2006.5 2006.75 2007 2007.25 2007.5 2007.75 ...
    2008 2008.25 2008.5 2008.75 2009 2009.25 2009.5 2009.75 2010 ...
    2010.25 2010.5 2010.75 2011 2011.25 2011.5, 2011.75, 2012], 'xticklabel', ...
    {'', '', 2000, '', '', ...
    '', 2001, '', '', '', 2002, '', '', '', 2003, '', '', '', 2004, '', ...
    '', '', 2005, '', '', '', 2006, '', '', '', 2007, '', '', '', 2008, ...
    '', '', '', 2009, '', '', '', 2010, '', '', '', 2011, '', '', '', 2012});

% turn the grid on
grid on

xlabel('Year');

% set the plot line thinkness to 1
set(gca, 'linewidth', 1);

hold off